152 load testing and stress simulation

bodiakof opened the pull request:
## How to Test

### Create Pending Commands  

`python manage.py create_pending_commands --host agent-0001`

You should see:

`Created 100 pending commands for host agent-0001`

and your DB will have 100 CommandHistory entries with status='pending'.


### Start Django in “test” mode


```
cd pydata_center
$Env:DJANGO_SETTINGS_MODULE="pydata_center.settings_test"
python manage.py runserver 0.0.0.0:8000
```

Runs on port 8000 with throttling disabled and extended JWT lifetime

### Start Toxiproxy

```
docker run -d --name toxiproxy \
  -p 8474:8474 \
  -p 9000:9000 \
  shopify/toxiproxy
```

8474 = control API
9000 = proxied service port

### Smoke‑test without network instability


```
cd pydata_center/monitoring/tests/performance
$env:LOCUST_TARGET="http://localhost:8000"; locust
```

Opens UI at http://localhost:8089
Run 10 users @ spawn rate 2
Verify agents fetch and finish the pending commands without errors.


### Test with network instability

```
cd pydata_center/monitoring/tests/performance
$env:ENABLE_TOXICS="true"; $env:LOCUST_TARGET="http://localhost:9000"; locust
```

Opens UI at http://localhost:8089
Run 50–100 users @ reasonable spawn rate

bodiakof committed:
feat(tests): add Locust load-testing script for agent–controller API

- Create `pydata_center/monitoring/tests/performance/locustfile.py`
  • Defines `AgentSimulator` HttpUser for simulating agent behavior

- Implements three core tasks with weighted scheduling:
  • `send_status` (weight: 3) – POSTs agent status to `/api/v1/server/status/`
  • `fetch_pending` (weight: 1) – GETs pending commands from `/api/v1/command/`
  • `submit_result` (weight: 1) – PATCHes command results to `/api/v1/command/submit-command/`

- Generates random hostnames (`agent-XXXX`) and IPs on start

- Configures `wait_time = between(1,3)` to mimic realistic agent intervals

- Provides single-line invocation for Windows PowerShell:
locust -f pydata_center/monitoring/tests/performance/locustfile.py --headless -u 50 -r 10 --host=http://localhost:8000

- Enables targeted scenarios via Locust’s `tasks` mapping

- Documents API_PREFIX override to match Django’s `settings.API_PREFIX`

This adds a repeatable, configurable load/stress testing harness for the controller’s HTTP endpoints.

bodiakof committed:
Merge branch 'develop' into 152-load-testing-and-stress-simulation

bodiakof committed:
added 'setting_test.py' to overide JWT and permissions from settings.py, some changes in 'locustfile.py'

bodiakof committed:
Add ALLOWED_HOSTS to settings.test

bodiakof committed:
Merge branch 'develop' into 152-load-testing-and-stress-simulation

bodiakof committed:
Remove 'DEFAULT_AUTHENTICATION_CLASSES' and 'DEFAULT_PERMISSION_CLASSES' from 'settings_test.py'

bodiakof committed:
Enhance AgentSimulator locust script with realistic status and result simulation

- Added `timestamp`, `os`, and `server_name` fields to the `/server/status/` payload for more realistic data
- Used `datetime.utcnow().isoformat()` for proper timestamp formatting
- Changed `submit_result` to send PATCH request to `/command/result/{id}/` instead of including `id` in payload
- Imported `datetime` module for timestamp generation

bodiakof committed:
Refactor locustfile: add realistic command ID handling and improve structure

- Replaced hardcoded random ID in `submit_result` with a predefined valid range
- Moved command ID submission into request body instead of URL path
- Initialized agent's hostname and IP in `on_start` with consistent formatting

bodiakof committed:
Merge branch 'develop' into 152-load-testing-and-stress-simulation

bodiakof committed:
feat(locustfile): add retry logic for JWT authentication

- Wrap initial token request in a 3‑attempt retry loop with 0.5s backoff
- Log each failed attempt as a warning
- Emit a single error and abort setup only after all retries fail
- Prevent `RuntimeError` from killing the user greenlet on intermittent network hiccups

bodiakof committed:
feat: add `create_pending_commands` management command

- Introduce `monitoring/management/commands/create_pending_commands.py` to simplify test setup by bulk‑creating pending `CommandHistory` entries.
- Allows specifying a target hostname (`--host`), number of commands (`--count`), and a shell snippet (`--shell`) for each command’s `params`.
- Uses Django’s `bulk_create` for efficient insertion of large command sets in one operation.
- Facilitates load testing and development workflows where controllers need pre‑seeded pending commands to dispatch.

bodiakof committed:
refactor(locustfile): improve auth resilience & dynamic command loading

- Fetch pending command IDs from `/api/v1/commands/?hostname=<host>&status=pending` instead of hard‑coding an ID range
- Store fetched IDs in `self.valid_ids`, append new pending commands in `fetch_pending()`, and remove them upon successful result submission
- Abort initialization gracefully if authentication fails, allowing other tasks to continue

bodiakof committed:
refactor: add agent activation guard to prevent task execution after failed login

- Introduced `self.active` flag to mark agents as fully initialized
- Ensures tasks like send_status, fetch_pending, and submit_result only run if setup succeeded
- Prevents unnecessary traffic or errors when token auth fails or backend is unreachable

bodiakof committed:
config: set ALLOWED_HOSTS to '*' to enable external access through Toxiproxy

bodiakof committed:
feat(locustfile): add toxiproxy support

- Integrated Toxiproxy for simulating latency and bandwidth limits
- Added logic to create proxy and inject/remove toxics automatically
- Controlled via env vars: ENABLE_TOXICS, TOXI_LATENCY, TOXI_JITTER, TOXI_RATE
- Added docstrings to all functions and the AgentSimulator class

mehalyna commented on pydata_center/monitoring/management/commands/create_pending_commands.py:
It would be helpful to validate that the `count` is a positive integer. Otherwise, `--count 0` or negative values will silently do nothing or cause confusion.

mehalyna commented on pydata_center/monitoring/management/commands/create_pending_commands.py:
For very large counts (e.g. 100,000), consider batching `bulk_create(..., batch_size=N)` to avoid DB limitations. Django supports this as a param.

mehalyna commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,270 @@
+            }
+        )
+        if resp.status_code == 200:
+            self.valid_ids.remove(cmd_id)
```
`submit_result()` could fail if a command ID is already handled by another agent. Add logic to detect and remove it:
```
if resp.status_code == 400:
    self.valid_ids.remove(cmd_id)
```

mehalyna commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,270 @@
+            logger.info(f'Proxy \'{PROXY_NAME}\' is available.')
+    except requests.exceptions.RequestException as e:
+        logger.error(f'Could not ensure proxy \'{PROXY_NAME}\': {e}')
+        raise SystemExit(1)
```
Check for invalid proxy configuration (listen/upstream mismatch) and optionally update if needed.

mehalyna commented on pydata_center/monitoring/tests/performance/locustfile.py:
Right now, tokens are only retrieved on `on_start`. If the token expires, the agent becomes unusable. Add a simple check:
```
if resp.status_code == 401:
    self.on_start() 
```

bodiakof committed:
feat: enforce positive count & batch insertion in create_pending_commands

- Validate `--count` is a positive integer; exit with error on zero or negative.
- Slice command list into chunks of 5000 and call `bulk_create(..., batch_size=5000)` to avoid DB limits.
- Preserve `--dry-run`, `--shell`, and `--type` flags for flexible usage.
- Add docstrings to class, `add_arguments()`, and `handle()` for clearer guidance.
- Pluralize success message (“1 command” vs “N commands”) for readability.

bodiakof committed:
fix(locust): handle 400 errors when submitting command results

- Avoid repeated attempts to submit results for commands that were already handled by other agents
- Log info and remove such command IDs from the pending list
- Wrap submission in try/except to catch and log unexpected errors

bodiakof committed:
feat(locust): add token auto-refresh and improve robustness of submit_result

- Introduced `_request_with_refresh` helper to retry requests if JWT expires (401).
- Replaced all HTTP calls in Locust tasks (`send_status`, `fetch_pending`, `submit_result`) with token-aware wrapper.
- Improved error handling in `submit_result`: removes command ID from list if already handled by another agent (400).

bodiakof committed:
fix(locustfile): recreate proxy if config mismatches expected listen/upstream

- Updates `ensure_proxy()` to verify existing proxy configuration.
- Deletes and recreates proxy if `listen` or `upstream` differs from expected.
- Ensures test environment has consistent proxy setup for Locust + Toxiproxy.
- Improves logging for visibility into proxy lifecycle.

bodiakof committed:
Add explicit 'name=' to all HTTP requests for clearer Locust stats

- Annotated all self.client.* calls with name= to avoid stat fragmentation
- Makes performance results easier to read and aggregate
- Matches each endpoint path exactly (e.g., /server/status/, /command/result/)

YelyzavetaV commented on pydata_center/monitoring/management/commands/create_pending_commands.py:
```diff
@@ -0,0 +1,95 @@
+        noun = 'command' if count == 1 else 'commands'
+        self.stdout.write(self.style.SUCCESS(
+            f'Created {count} pending {noun} for host {hostname}'
+        ))
```
I'm not a fan of hardcoded batch size, especially that use the naming convention meant for global constants. Perhaps, make into another command-line argument set in `add_arguments` with the default of 5000?

YelyzavetaV commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,326 @@
+    if (ENABLE_TOXICS and environment.host and environment.host !=
+            'http://localhost:8000'):
+        remove_toxics()
+        time.sleep(2)
```
Perhaps, you could parametrize sleeping interval using an environmental variable?

YelyzavetaV commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,326 @@
+            f'{API_PREFIX}/commands/',
+            name='/commands/',
+            params={'hostname': self.hostname, 'status': 'pending'}
+        )
```
Could be a good idea to handle exceptions on failed HTTP requests using `try/except`.

YelyzavetaV commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,326 @@
+                'os': random.choice(['Linux', 'Windows', 'macOS']),
+                'server_name': self.hostname.replace('-', '_'),
+            }
+        )
```
Same comment about handling exceptions on failed HTTP requests.

YelyzavetaV commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,326 @@
+            logger.warning('Token expired; refreshing and retrying')
+            self.on_start()  # Re-authenticate
+            response = method(*args, **kwargs)
+        return response
```
You only handle a failure once and if the second attempt fails, you return a failed response. Consider parametrizing the number of retries and always logging a failure no matter which attempt it occurs on.

YelyzavetaV commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,326 @@
+        if pending.status_code == 200:
+            self.valid_ids = [cmd['id'] for cmd in pending.json()]
+        else:
+            self.valid_ids = []
```
If Locust is run using multiple threads, modifying `self.valid_ids` may cause a racing condition. Consider using `Queue.Queue` for thread-safety.

bodiakof replied:
Here's what ai says: _"Locust spawns multiple processes (or greenlets via gevent) — not threads — for its users. Each User class instance runs in isolation, so race conditions on instance attributes like self.valid_ids don’t happen."_

YelyzavetaV commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,326 @@
+LATENCY_MS = int(os.getenv('TOXI_LATENCY', '500'))
+JITTER_MS = int(os.getenv('TOXI_JITTER', '100'))
+RATE_LIMIT = int(os.getenv('TOXI_RATE', '80000'))
+ENABLE_TOXICS = os.getenv('ENABLE_TOXICS', 'true').lower() == 'true'
```
Add a module-level docstring documenting all the env variables.

YelyzavetaV commented on pydata_center/monitoring/tests/performance/locustfile.py:
```diff
@@ -0,0 +1,326 @@
+
+API_PREFIX = '/api/v1'
+USERNAME = 'locust_tester'
+PASSWORD = 'supersecret'
```
These shouldn't be hardcoded.

bodiakof committed:
feat(create_pending_commands): add --batch-size argument to customize bulk_create batching

- Replaced hardcoded BATCH_SIZE = 5000 with dynamic --batch-size CLI option
- Default remains 5000 for backwards compatibility
- Added validation to ensure --batch-size >= 1

bodiakof committed:
Parametrize toxic cleanup delay with TOXI_CLEANUP_DELAY env variable

- Replaced hardcoded `time.sleep(2)` with a configurable delay
- Introduced `TOXIC_CLEANUP_DELAY` env var (default: 2 seconds)

bodiakof committed:
Add error handling and config overrides to Locust test script

- Wrapped login, token parsing, and initial command fetch in try/except blocks
- Wrapped /server/status/ POST in try/except to avoid test interruptions

bodiakof committed:
Improve token refresh logic with configurable retries and better logging

- Updated _request_with_refresh() to support multiple retry attempts (configurable via MAX_REFRESH_RETRIES env var)
- Log every failed attempt and final failure if all retries are exhausted

bodiakof committed:
fix Toxiproxy re-creation on every test run

- Updated expected_listens to accept both '0.0.0.0:9000' and '[::]:9000'
- Only recreate proxy if config truly mismatches (upstream or listen)
- Improved logging to show actual config mismatch

Prevents unnecessary proxy teardown/recreate loops and reduces login failures under load.

bodiakof committed:
use os.getenv to make username and password environment variables

bodiakof committed:
add module-level docstring for env vars and group config constants

- Document all supported environment variables, their purpose, and default values
- Group related configuration constants (auth, toxiproxy, retry) for readability
- Allow overriding TOXIPROXY_API and PROXY_NAME via environment
- Use os.getenv() consistently for config overrides

bodiakof committed:
add pytest-benchmark tests for core API endpoints

- Implement performance benchmarks for authentication token issuance, fetching pending commands, sending status updates, fetching commands, and submitting command results.
- Validate responses and assert expected HTTP status codes.
- Simulate realistic payloads for status updates with randomized uptime and health.
- Prepare groundwork for automated performance regression tracking and monitoring under load.

bodiakof committed:
monitoring tests fix

bodiakof committed:
Merge branch 'develop' into 152-load-testing-and-stress-simulation

bodiakof committed:
pre-comit fix

bodiakof committed:
monitoring-test fix

bodiakof committed:
monitoring-test fix

bodiakof committed:
add pyproject.toml to unify isort and yapf formatting rules

Ensures consistent import sorting and code formatting across all environments,
reducing diffs and preventing merge conflicts caused by styling differences.

bodiakof committed:
add pyproject.toml to unify isort and yapf formatting rules

Ensures consistent import sorting and code formatting across all environments,
reducing diffs and preventing merge conflicts caused by styling differences.

bodiakof committed:
motitoring-tests fix (plese work this time!)

bodiakof committed:
monitoring-tests fix (maybe this time?)

bodiakof committed:
monitoring-tests fix

bodiakof committed:
Merge branch 'develop' into 152-load-testing-and-stress-simulation

bodiakof committed:
Merge branch 'develop' into 152-load-testing-and-stress-simulation

bodiakof committed:
monitoring-tests fix

bodiakof committed:
monitoring-tests fix

bodiakof committed:
monitoring-tests fix

bodiakof committed:
monitoring-tests fix

bodiakof committed:
monitoring-tests fix

bodiakof committed:
pre-commit fix

bodiakof committed:
monitoring-tests fix

bodiakof committed:
motitoring-tests fix

bodiakof committed:
monitoring-tests fix

bodiakof committed:
Merge branch '152-load-testing-and-stress-simulation' of https://github.com/ITA-Dnipro/PyDataCenter5000 into 152-load-testing-and-stress-simulation

bodiakof committed:
debug the environment

bodiakof committed:
monitoring-tests fix

bodiakof committed:
monitoring-tests fix

bodiakof committed:
fix

bodiakof committed:
Merge branch 'develop' into 152-load-testing-and-stress-simulation

bodiakof committed:
remove 4th (bohdan-monitoring-tests.yml) test workflow

bodiakof committed:
Remove bohdan-monitoring-tests workflow
